cmd/binlog-collector/controller/controller.go

```diff
@@ -0,0 +1,196 @@
+package controller
+
+import (
+	"bytes"
+	"os/exec"
+	"strings"
+
+	"github.com/pkg/errors"
+
+	"github.com/percona/percona-xtradb-cluster-operator/cmd/binlog-collector/db"
+	"github.com/percona/percona-xtradb-cluster-operator/cmd/binlog-collector/storage"
+)
+
+type Controller struct {
+	dbm            *db.Manager
+	sm             *storage.Manager
+	lastSet        string
+	pxcServiceName string
+	pxcUser        string
+	pxcPass        string
+}
+
+type Config struct {
+	PXCServiceName string
+	PXCUser        string
+	PXCPass        string
+	S3Endpoint     string
+	S3accessKeyID  string
+	S3accessKey    string
+	S3bucketName   string
+}
+
+func New(c Config) (*Controller, error) {
+	sm, err := storage.NewManager(c.S3Endpoint, c.S3accessKeyID, c.S3accessKey, c.S3bucketName, "last-set.txt", true)
+	if err != nil {
+		return nil, errors.Wrap(err, "new storage manager")
+	}
+
+	// get last binlog set stored on S3
+	lastSet, err := sm.GetObjectContent(sm.LastSetObjectName)
```

Comment:
All this faffing around `sm.LastSetObjectName` doesn't look alright to me. You set it via New... method but then straight away access it via the public field of the object. 
Anyway, **storage**  as the abstraction that provides access to the store shouldn't know anything about files naming conventions or its purpose etc. Otherwise what a reason for having it as the separate package?
Moreover, look at usage patterns - you always LastSetObjectName to the storage methods from this package. So it should belong to this package. Make it const (since it is a const in fact) at this package and use it.

And by the way, LastSetObjectName sounds like it's a name of some object when in fact it is a name of the file on the object store (which doesn't have to be necessary an object store - it could be some kind of block storage at some point of time) which contains the last GTID.

Reply:
https://github.com/percona/percona-xtradb-cluster-operator/pull/622/commits/cec0c704c6c8cb2eb0957e2123bd8b04478b8605